File: java-demo/src/main/java/demo3/Ex2.java
package demo3;

import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import akka.stream.ClosedShape;
import akka.stream.FanInShape2;
import akka.stream.Graph;
import akka.stream.Inlet;
import akka.stream.Outlet;
import akka.stream.UniformFanInShape;
import akka.stream.javadsl.FlowGraph;
import akka.stream.javadsl.RunnableGraph;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import akka.stream.javadsl.ZipWith;
import scala.concurrent.Await;
import scala.concurrent.Future;
import scala.concurrent.duration.Duration;
import scala.runtime.BoxedUnit;

public class Ex2 {
  public static void main(String[] args) throws Exception {
    ActorSystem system = ActorSystem.create("demo3");
    ActorMaterializer materializer = ActorMaterializer.create(system);

    Graph<FanInShape2<Integer, Integer, Integer>, BoxedUnit> zip =
    		  ZipWith.create((Integer left, Integer right) -> Math.max(left, right));
    // Create a new fan-in junction that takes 3 Integer inputs and outputs the max
    Graph<UniformFanInShape<Integer, Integer>, BoxedUnit> maxOfThree = 
      FlowGraph.create( b -> {
        FanInShape2<Integer, Integer, Integer> zip1 = b.add(zip);
        FanInShape2<Integer, Integer, Integer> zip2 = b.add(zip);

        b.from(zip1.out()).toInlet(zip2.in0());

        // return the shape, which has three inputs and one output
        return new UniformFanInShape<Integer, Integer>(zip2.out(), 
          new Inlet[] {zip1.in0(), zip1.in1(), zip2.in1()});
      });

    Sink<Integer, Future<Integer>> sink = Sink.head();

    RunnableGraph<Future<Integer>> runnable = RunnableGraph.<Future<Integer>>fromGraph(
      FlowGraph.create(sink, (b, out) -> {
        Outlet<Integer> s1 = b.add(Source.single(1)).outlet();
        Outlet<Integer> s2 = b.add(Source.single(2)).outlet();
        Outlet<Integer> s3 = b.add(Source.single(3)).outlet();
        UniformFanInShape<Integer, Integer> pm3 = b.add(maxOfThree);

        b.from(s1).toInlet(pm3.in(0));
        b.from(s2).toInlet(pm3.in(1));
        b.from(s3).toInlet(pm3.in(2));
        b.from(pm3.out()).to(out);
        return ClosedShape.getInstance();
      }
    ));

    Future<Integer> max = runnable.run(materializer);
    Integer res = Await.result(max, Duration.Inf());
    System.out.println(res);
  }
}

File: scala-demo/src/main/scala/demo3/Ex2.scala
package demo3

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.stream._
import akka.stream.scaladsl._
import scala.concurrent.duration._
import akka.stream.scaladsl.FlowGraph.Implicits._
import FlowGraph.Implicits._
import scala.concurrent.Future
import scala.concurrent.Await

object Ex2 extends App {
  implicit val system = ActorSystem("demo3")
  implicit val materializer = ActorMaterializer()
  import system.dispatcher

  // Create a new fan-in junction that takes 3 Int inputs and outputs the max
  val maxOfThree = FlowGraph.partial() { implicit b =>
    val zip1 = b.add(ZipWith[Int, Int, Int](math.max))
    val zip2 = b.add(ZipWith[Int, Int, Int](math.max))

    zip1.out ~> zip2.in0

    UniformFanInShape(zip2.out, zip1.in0, zip1.in1, zip2.in1)
  }

  // and now let's create a stream using this new fan-in shape

  val sink = Sink.head[Int]
  val g = FlowGraph.closed(sink) { implicit b =>
    out =>
      val s1 = Source.single(1)
      val s2 = Source.single(2)
      val s3 = Source.single(3)
      val pm3 = b.add(maxOfThree)

      s1 ~> pm3
      s2 ~> pm3
      s3 ~> pm3
      pm3 ~> out
  }
  val max = g.run()
  val res = Await.result(max, Duration.Inf)
  println(res)
}